gcovr/json_generator.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding:utf-8 -*-
+
+# This file is part of gcovr <http://gcovr.com/>.
+#
+# Copyright 2019 the gcovr authors
+# This software is distributed under the BSD license.
+
+import json
+import os
+import sys
+import functools
+
+from .utils import presentable_filename, Logger
+from .coverage import FileCoverage
+
+
+JSON_FORMAT_VERSION = 0.1
+PRETTY_JSON_INDENT = 4
+
+
+#
+# Produce gcovr JSON report
+#
+def print_json_report(covdata, output_file, options):
+    r"""produce an JSON report in the format partially
+    compatible with gcov JSON output"""
+
+    gcovr_json_root = {}
+    gcovr_json_root['gcovr/format_version'] = JSON_FORMAT_VERSION
+    gcovr_json_root['files'] = []
+
+    for no in sorted(covdata):
+        gcovr_json_file = {}
+        gcovr_json_file['file'] = presentable_filename(covdata[no].filename, root_filter=options.root_filter)
+        gcovr_json_file['lines'] = _json_from_lines(covdata[no].lines)
+        gcovr_json_root['files'].append(gcovr_json_file)
+
+    write_json = json.dump
+    if options.prettyjson:
+        write_json = functools.partial(write_json, indent=PRETTY_JSON_INDENT)
+
+    if options.output:
+        with open(options.output, 'w') as output:
+            write_json(gcovr_json_root, output)
+    else:
+        write_json(gcovr_json_root, sys.stdout)
+
+
+#
+#  Get coverage from already existing gcovr JSON files
+#
+def gcovr_json_files_to_coverage(filenames, covdata, options):
+    r"""merge a coverage from multiple reports in the format
+    partially compatible with gcov JSON output"""
+
+    logger = Logger(options.verbose)
+
+    for filename in filenames:
+        gcovr_json_data = {}
+        logger.verbose_msg("Processing JSON file: {}", filename)
+
+        with open(filename, 'r') as json_file:
+            gcovr_json_data = json.load(json_file)
+
+        assert gcovr_json_data['gcovr/format_version'] == JSON_FORMAT_VERSION
+
+        coverage = {}
+        for gcovr_file in gcovr_json_data['files']:
+            file_path = os.path.join(os.path.abspath(options.root), gcovr_file['file'])
+            file_coverage = FileCoverage(file_path)
+            _lines_from_json(file_coverage, gcovr_file['lines'])
+            coverage[file_path] = file_coverage
+
+        _split_coverage_results(covdata, coverage)
+
+
+def _split_coverage_results(covdata, coverages):
+    for coverage in coverages.values():
+        if coverage.filename not in covdata:
+            covdata[coverage.filename] = FileCoverage(coverage.filename)
+
+        covdata[coverage.filename].update(coverage)
+
+
+def _json_from_lines(lines):
+    json_lines = [_json_from_line(lines[no]) for no in sorted(lines)]
+    return json_lines
+
+
+def _json_from_line(line):
+    json_line = {}
+    json_line['branches'] = _json_from_branches(line.branches)
+    json_line['count'] = line.count
+    json_line['line_number'] = line.lineno
+    json_line['unexecuted_block'] = line.count == 0
```

Comment:
I'm not sure what the correct semantics for the `unexecuted_block` field are. Gcovr does not currently parse per-block data and cannot confidently say whether a line contains unexecuted blocks. Blocks have to do with partial coverage: a line can be covered but still contain unexecuted blocks, e.g. `if (foo()) bar();` where `foo()` always returns false.

I would get rid of this field in order to avoid incompatibilities with the gcov JSON format.

When loading, this field is used to populate the `noncode` property. That's a good idea, so maybe it would be best explicitly to store a `gcovr/noncode` field instead?

Reply:
How about just to remove `unexecuted_block` filed (or `gcovr/noncode`after renaming)?
During conversion JSON file to `covdata` I can update the `noncode` property after fetching the `lineno`.

```diff
--- a/gcovr/json_generator.py
+++ b/gcovr/json_generator.py
@@ -92,7 +92,6 @@ def _json_from_line(line):
     json_line['branches'] = _json_from_branches(line.branches)
     json_line['count'] = line.count
     json_line['line_number'] = line.lineno
-    json_line['unexecuted_block'] = line.count == 0
     return json_line
 
 
@@ -114,8 +113,8 @@ def _lines_from_json(file, json_lines):
 
 
 def _line_from_json(line, json_line):
-    line.noncode = json_line['unexecuted_block']
     line.count = json_line['count']
+    line.noncode = line.count == 0
     _branches_from_json(line, json_line['branches'])
```

Reply:
No, noncode means that a line should not be included in the coverage calculation. That's not the same as an uncovered line. The two alternatives are:

* do not output noncode lines to the JSON. This is correct, but lossy.
* change `unexecuted_block` to something like `gcovr/noncode` or `gcovr/excluded`.

I'll leave it to you to determine the better approach.

Reply:
I think, now it should be correct. `gcovr/noncode` filed has `line.noncode` value.